File: ios/ios/ViewController.m
//
//  ViewController.m
//  ios
//
//  Created by Kevin Galligan on 2/23/16.
//  Copyright © 2016 Kevin Galligan. All rights reserved.
//

#import "ViewController.h"
#import "co/touchlab/droidconandroid/presenter/ConferenceDataPresenter.h"
#import "co/touchlab/droidconandroid/presenter/ConferenceDayHolder.h"
#import "co/touchlab/droidconandroid/presenter/AppManager.h"
#import "co/touchlab/droidconandroid/data/Block.h"
#import "co/touchlab/droidconandroid/data/Venue.h"
#import "android/content/IOSContext.h"
#import "PlatformContext_iOS.h"
#import "java/util/ArrayList.h"
#import "java/util/List.h"
#import "NoteTableViewCell.h"


@interface ViewController ()

@property (nonatomic, assign) long track;
@property (nonatomic, strong) NSMutableArray *notesArray;
@property (nonatomic, strong) NSMutableArray *imagesArray;
@property (nonatomic, strong) PlatformContext_iOS *platformContext;
@property (nonatomic, strong) DCPConferenceDataPresenter *dataPresenter;
@property (nonatomic, strong) JavaUtilArrayList *notes;
@end

BOOL allEvents = NO;

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (self.tabBarController.selectedIndex == 0) {
        allEvents = YES;
        self.navigationItem.title = @"Droidcon NYC";
    } else {
        allEvents = NO;
        self.navigationItem.title = @"My Agenda";
    }

    // Hide the nav bar shadow
    [self.navigationController.navigationBar setShadowImage:[[UIImage alloc] init]];
    [self.navigationController.navigationBar setBackgroundImage:[[UIImage alloc]init] forBarMetrics:UIBarMetricsDefault];
    
    self.navigationController.navigationBar.translucent = false;
}

- (void) viewWillAppear:(BOOL)animated
{
    // refresh every time it appears so that we see updates, doesn't seem to affect scroll position of list
    [self loadConferenceSchedule];
    self.tableView.tableHeaderView = nil;
    self.tableView.tableFooterView = nil;
    
    self.tableView.delegate = self.platformContext;
    self.tableView.dataSource = self.platformContext;
    
    // will refresh data from server only if it is old
    [self.dataPresenter refreshFromServer];
}

- (void)createSDASimple
{
    if(self.platformContext == nil)
    {
    PlatformContext_iOS *pcios = [PlatformContext_iOS new];
    self.platformContext = pcios;
    self.platformContext.reloadDelegate = self;
    self.dataPresenter = [[DCPConferenceDataPresenter alloc] initWithAndroidContentContext:[DCPAppManager getContext] withDCPConferenceDataHost:pcios withBoolean:allEvents];
    }
    else
    {
        [self.dataPresenter refreshConferenceData];
    }
}

- (IBAction)updateTable:(id)sender
{
    if ([self.dayChooser selectedSegmentIndex] == 0) {
        self.platformContext.isDayTwo = NO;
    } else {
        self.platformContext.isDayTwo = YES;
    }
    [self.platformContext updateTableData];
    [self.tableView reloadData];
}

- (void)loadImageWithPath:(NSString *)imagePath
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory,
                                                         NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    NSString *path = [documentsDirectory stringByAppendingPathComponent:imagePath];
    UIImage *image = [UIImage imageWithContentsOfFile:path];
    UIImageView *imgView = [[UIImageView alloc] initWithImage:image];
    [self.imagesArray addObject:imgView];
}

#pragma PlatformContext_iOS - Delegate
- (void)reloadTableView
{
    [self.tableView reloadData];
}

- (void)showEventDetailViewWithEvent:(DCDEvent *)event andIndex:(long)index
{
    self.track = index;
    [self performSegueWithIdentifier:@"ShowEventDetail" sender:event];
}

- (void)showBlockDetailViewWithBlock:(DCDBlock *)block
{
    [self performSegueWithIdentifier:@"ShowBlockDetail" sender:block];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{

    if ([[segue identifier] isEqualToString:@"ShowEventDetail"]) {
        ShowEventDetailViewController *detailVC = segue.destinationViewController;
        DCDEvent *event = (DCDEvent *)sender;
        NSArray *speakers = [self.platformContext getSpeakersArrayFromEvent:event];
        detailVC.titleString = [event->name_ stringByReplacingOccurrencesOfString:@"Android" withString:@"[Sad Puppy]"];
        detailVC.descriptionString = [event->description__ stringByReplacingOccurrencesOfString:@"Android" withString:@"[Sad Puppy]"];
        detailVC.event = event;
        detailVC.speakers = speakers;
        detailVC.dateTime = [self.platformContext getEventTimeFromStart:[event getStartFormatted] andEnd:[event getEndFormatted]];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)loadConferenceSchedule
{
    [self createSDASimple];
}

@end

File: ios/ios/ViewController.swift
//
//  ViewController.swift
//  ios
//
//  Created by Kachi Nwaobasi on 7/11/17.
//  Copyright © 2017 Kevin Galligan. All rights reserved.
//

import Foundation
import JRE
import UIKit

class ViewController : UIViewController {

    var track: Int!
    var notesArray: [String]!
    var imagesArray: [UIImageView]!
    var platformContext: PlatformContext_iOS!
    var conferencePresenter: DPRESConferenceDataViewModel!
    var schedulePresenter: DPRESScheduleDataViewModel!
    var notes: JavaUtilArrayList!
    var allEvents = false
    @IBOutlet weak var dayChooser: UISegmentedControl!
    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        allEvents = self.tabBarController?.selectedIndex == 0
        navigationItem.title = allEvents ? "Droidcon NYC" : "My Agenda"
        
        // Hide the nav bar shadow
        navigationController?.navigationBar.shadowImage = UIImage()
        navigationController?.navigationBar.setBackgroundImage(UIImage(), for: UIBarMetrics.default)
        navigationController?.navigationBar.isTranslucent = false
        
        tableView.estimatedRowHeight = 44
        tableView.rowHeight = UITableViewAutomaticDimension
    }
    
    override func viewWillAppear(_ animated: Bool) {
        // refresh every time it appears so that we see updates, doesn't seem to affect scroll position of list
        loadConferenceSchedule()
        
        tableView.delegate = platformContext
        tableView.dataSource = platformContext
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "ShowEventDetail" {
            let detailViewController = segue.destination as! ShowEventDetailViewController
            let networkEvent = sender as! DDATEvent
            let speakers = platformContext.getSpeakersArray(from: networkEvent) as! [DDATEventSpeaker]
            detailViewController.titleString = networkEvent.getName().replacingOccurrences(of: "Android", with: "[Sad Puppy]")
            detailViewController.descriptionString = networkEvent.getDescription().replacingOccurrences(of: "Android", with: "[Sad Puppy]")
            detailViewController.event = networkEvent
            //detailViewController.speakers = speakers
            detailViewController.dateTime = platformContext.getEventTime(startTime: networkEvent.getStartFormatted()! as NSString, andEnd: networkEvent.getEndFormatted()! as NSString)
        }
    }
    
    private func createSDASimple() {
        if platformContext == nil {
            platformContext = PlatformContext_iOS()
            platformContext.reloadDelegate = self
            conferencePresenter = DPRESConferenceDataViewModel.forIosWithBoolean(allEvents)
            schedulePresenter = DPRESScheduleDataViewModel.forIos()
            schedulePresenter.register__(with: platformContext, withBoolean: true)
            
        } 
    }
    
    @IBAction func updateTable(_ sender: AnyObject) {
        platformContext.isDayTwo = dayChooser.selectedSegmentIndex > 0
        platformContext.updateTableData()
        tableView.reloadData()
    }
    
    func loadImage(with imagePath: String) {
        let paths = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true)
        let documentsDirectory = URL(fileURLWithPath: paths[0])
        let path = documentsDirectory.appendingPathComponent(imagePath)
        let image = UIImage(contentsOfFile: path.absoluteString)
        let imageView = UIImageView(image: image)
        imagesArray.append(imageView)
    }
    
    private func loadConferenceSchedule() {
        createSDASimple()
    }
    
}

//MARK: PlatformContext_iOSDelegate
extension ViewController : PlatformContext_iOSDelegate {
    
    func reloadTableView() {
        tableView.reloadData()
    }
    
    func showEventDetailView(with networkEvent: DDATTimeBlock, andIndex index: Int) {
        track = index
        performSegue(withIdentifier: "ShowEventDetail", sender: networkEvent)
    }
    
    func showBlockDetailView(with networkBlock: DDATBlock) {
        performSegue(withIdentifier: "ShowBlockDetail", sender: networkBlock)
    }
}
